### ISSUE TITLE
Report to admin if statements fail sending to LRS

### ISSUE THREAD

**Issue by garemoko** [davidpesce/moodle-logstore_xapi#37]
**Motive**
1. It's a nice thing to have
   @ryansmith94 comments:
   
   > I think it might be a good idea to notify someone if an event has been waiting more than X minutes or retried more than Y times. This is useful if:
   > - The LRS is down for X minutes and the plugin user doesn't know
   > - A statement is invalid (hence the plugin has a bug)
   > - The LRS is incorrectly returning an error for a valid statement (hence the LRS has a bug)
2. It's a step towards ensuring it completely replaces the quiz specific module
   See https://github.com/leo-platforms/moodle-quiz-tincan
3. (against) It's out of scope (via @jlowe64) 
   The LRS or some other monitoring system should report if there's a problem

**Result**
- Configuration for the plugin admin to specify alert behavior and email
- Log errors and monitor number of errors or length of unavailability
- Send the alert

See https://github.com/leo-platforms/moodle-quiz-tincan/tree/master/local/resendtincandata

**Comment by jlowe64** [comment]
@ryansmith94 totally agree with your opinion on the sending forever thing, and yeah would depend on how the retries are handled. 

I would say the "retry into infinity" problem needs to be addressed so that we may suspend normal crons until a success is managed. Cause if you have 3000 statements a day, and you have the cron set to every minute, and your LRS is down, you may run into a massive lag as the DB gets filled with more and more statements.

**Comment by ryasmi** [comment]
Yeah @jlowe64 "retry to infinity" needs to be addressed and I think that needs to be done first. Perhaps after X attempts it stops retrying and alerts (and X could be configured).

**Comment by jlowe64** [comment]
Ben at HT2 has stated that this may have been addressed in a version of JISC's plugin. Will wait to confirm with Ben that this has happened so that we do not duplicate work.

**Comment by garemoko** [comment]
@jlowe64 what is the JISC plugin? Is that a fork of this or something separate?

**Comment by jlowe64** [comment]
JISC is one of the organisations that are helping with development of the plugin, they are also using it. I believe they have a fork somewhere, but I am unsure if it's public. I need more info from Ben at HT2 before I know more.

**Comment by jlowe64** [comment]
@garemoko

**Comment by ryasmi** [comment]
@michaelwebbjisc could you provide any info on [my earlier comment](#issuecomment-171401582) and this issue in general?

**Comment by nadavkav** [comment]
Yes please! (I already need this, as I am trying to implement missing events classes for a new H5P plugin)

**Comment by garemoko** [comment]
I'm leaning more on the "The LRS or some other monitoring system should report if there's a problem" side of this argument now. I do think we should do a better job of retrying, but alerts are less of a priority for me.

**Comment by ryasmi** [comment]
Closing this, maybe we can raise a new issue for preventing infinite retries @davidpesce and discuss a solution in that issue?
